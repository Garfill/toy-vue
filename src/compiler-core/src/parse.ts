import { NodeTypes } from "./ast"

const openDelimiter = '{{'
const closeDelimiter = '}}'

const enum TagType {
  OPEN,
  END,
}

export function baseParse(content: string) {
  const context = createContext(content)
  return createRoot(parseChildren(context))
}

function parseChildren(context) {
  const nodes: any[] = []
  let node
  let s = context.source
  if (s.startsWith(openDelimiter)) {
    node = parseInterpolation(context)
  } else if (s[0] === '<') {
    if (/[a-z]/i.test(s[1])) {
      node = parseElement(context)
    }
  }

  if (!node) {
    // 默认当成普通文本
    node = parseText(context, context.source.length)
  }

  nodes.push(node)
  return nodes
}

function createContext(content: string) {
  return {
    source: content
  }
}

function createRoot(children) {
  return {
    children
  }
}


function parseInterpolation(context) {
  // {{message}} -> message
  const closeIndex = context.source.indexOf('}}')
  // context.source = context.source.slice(openDelimiter.length)
  advanceBy(context, openDelimiter.length)

  // 获取内部文本长度
  const contentLength = closeIndex - openDelimiter.length
  // const rawContent = context.source.slice(0, contentLength)
  // // context.source = context.source.slice(contentLength + closeDelimiter.length)
  // advanceBy(context, contentLength + closeDelimiter.length)
  const rawContent = parseTextData(context, contentLength)
  const content = rawContent.trim()
  advanceBy(context, closeDelimiter.length)
  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    }
  }
}

function advanceBy(context, step: number) {
  context.source = context.source.slice(step)
}


function parseElement(context: any): any {
  // <div></div>
  // 开端标签
  const element = parseTag(context, TagType.OPEN)
  // 闭合标签
  parseTag(context, TagType.END)

  console.log('>>>>>>>>>>>', context.source)
  return element
}

export function parseTag(context, type: TagType) {
  let match: any = /^<\/?([a-z]*)/i.exec(context.source)
  let tag = match[1]
  advanceBy(context, match[0].length)
  advanceBy(context, 1)

  if (type === TagType.END) return

  return {
    type: NodeTypes.ELEMENT,
    tag
  }
}

function parseText(context: any, length: number) {
  const content = parseTextData(context, length)
  console.log(context.source.length === 0)
  return {
    type: NodeTypes.TEXT,
    content,
  }
}

function parseTextData(context, length) {
  let content = context.source.slice(0, length)
  advanceBy(context, content.length)
  return content
}

