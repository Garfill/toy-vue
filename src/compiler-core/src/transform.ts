import { NodeTypes } from "./ast";
import { helperMapName, TO_DISPLAY_STRING } from "./runtimeHelpers";

export default function transform(root, options = {}) {
  const context = createTransformContext(root, options)
  // 1. DFS 遍历
  // 2. 修改对应节点
  traverseNode(root, context)
  createRootCodegen(root, context);

  root.helpers = [...context.helpers.keys()]
}

function createRootCodegen(root, context) {
  root.codegenNode = root.children[0];
}

function traverseNode(node: any, context) {
  console.log('traverse >>>>', node)
  const nodeTransform = context.nodeTransform || []
  for (let i = 0; i < nodeTransform.length; i++) {
    const transform = nodeTransform[i];
    transform(node)
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(helperMapName[TO_DISPLAY_STRING])
      break
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChildren(node, context);
      break;
    default:
      break
  }
}

function traverseChildren(node: any, context: any) {
  const { children } = node;
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    traverseNode(child, context);
  }
}

function createTransformContext(root: any, options: any) {
  const context = {
    root,
    nodeTransform: options.nodeTransform,
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1)
    }
  }

  return context
}

