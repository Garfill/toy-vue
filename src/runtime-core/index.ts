export * from './h'
export * from './helpers/renderSlot'
export { getCurrentInstance, registerRuntimeCompiler } from './component'
export { createTextVNode, createElementVNode } from './vnode'
export { provide, inject } from './apiInject'
export { createRenderer } from './renderer'
export { nextTick } from './scheduler'
export { toDisplayString } from '../share'
export * from '../reactive'