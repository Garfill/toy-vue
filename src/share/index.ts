export const extend = Object.assign

export function isObject(value) {
  return value !== null && typeof value === 'object'
}

export function hasChange(old, value) {
  return !Object.is(old, value)
}

export function isFunction(value: any)  {
  return typeof value === 'function'
}

export function isString(value) {
  return typeof value === 'string' || typeof value === 'number'
}

export function isArray(value) {
  return Array.isArray(value)
}

export function hasOwn(value, key) {
  return Object.prototype.hasOwnProperty.call(value, key)
}

export function toHandlerKey(event: string) {
  return "on" + capitalize(event)
}

export function capitalize(string: string) {
  return string.charAt(0).toUpperCase() + string.slice(1)
}


const camelizeRE = /-(\w)/g
export function camelize(string: string) {
  return string.replace(camelizeRE, (_, c) => (c ? c.toUpperCase() : ''))
}

export const isOn = (key: string) => /^on[A-Z]/.test(key)

export const EMPTY_OBJ = {}