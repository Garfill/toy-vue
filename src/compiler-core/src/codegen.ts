import { isString } from "../../share"
import { NodeTypes } from "./ast"
import { CREATE_ELEMENT_VNODE, helperMapName, TO_DISPLAY_STRING } from "./runtimeHelpers"

export function generate(ast) {
  const context = createGenerateContext(ast)
  const { push } = context

  genFunctionPreamble(ast, context)

  push('return ')
  const functionName = 'rendner'
  const args = ['_ctx', '_cache']
  const signature = args.join(', ')
  push(`function ${functionName}(${signature}) {`)

  // 函数内容
  push('return ')
  genNode(ast.codegenNode, context)


  push("}")
  return {
    code: context.code
  }
}

function genFunctionPreamble(ast: any, context) {
  const { push } = context
  const VueBinding = 'Vue'
  const helpers = ast.helpers
  const aliasHelper = (s) => `${helperMapName[s]}: _${helperMapName[s]}`
  if (helpers.length) {
    push(`const {${helpers.map(aliasHelper).join(', ')}} = ${VueBinding}`)
  }
  push('\n')
}

function genNode(node, context) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genTextCode(node, context)
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolationCode(node, context)
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpressionCode(node, context)
      break;
    case NodeTypes.ELEMENT:
      genElementCode(node, context)
      break;
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpressionCode(node, context)
      break;
    default:
      break;
  }
}

function createGenerateContext(ast: any) {
  let context = {
    ast,
    code: '',
    push(source) {
      context.code += source
    },
    helper(key) {
      return `_${helperMapName[key]}`
    }
  }

  return context
}

function genTextCode(node: any, context) {
  const { push } = context
  push(`'${node.content}'`)
}

function genInterpolationCode(node: any, context: any) {
  const { push, helper } = context
  push(`${helper(TO_DISPLAY_STRING)}(`)
  genNode(node.content, context)
  push(')')
}

function genExpressionCode(node: any, context: any) {
  const { push } = context
  push(node.content)
}

function genElementCode(node: any, context: any) {
  const { push, helper } = context
  const { tag, children, props } = node
  // 此处的 node ，参照transformElement的vnodeElement，其中的children就是原来的children[0]
  push(`${helper(CREATE_ELEMENT_VNODE)}(`)
  // for (let i = 0; i < children.length; i++) {
  //   const child = children[i];
  //   genNode(child, context)
  // }
  genNodeList(genNullable([tag, props, children]), context)
  // genNode(children, context)
  push(')')
}

function genCompoundExpressionCode(node: any, context: any) {
  const { push } = context
  const { children } = node
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (isString(child)) {
      push(child)
    } else {
      genNode(child, context)
    }
  }
}

function genNullable(args) {
  return args.map((arg) => arg ? arg : "null")
}

function genNodeList(nodes, context) {
  console.log('nodes>>>>>>>', nodes)
  const { push } = context
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (isString(node)) {
      push(node)
    } else {
      genNode(node, context)
    }
    if (i < nodes.length - 1) {
      push(', ')
    }
  }
}

