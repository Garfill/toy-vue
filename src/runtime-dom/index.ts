import { createRenderer } from "../runtime-core"
import { isOn } from "../share/index"

function createElement(type) {
  return document.createElement(type)
}

function patchProps(el, key, prevVal, nextVal) {
  if (isOn(key)) {
    const event = key.slice(2).toLowerCase() // 取出事件名
    el.addEventListener(event, nextVal)
  } else {
    // 当新的props是 undefined/null ，直接删除属性
    if (nextVal === undefined || nextVal === null) {
      el.removeAttribute(key)
    } else {
      el.setAttribute(key, nextVal)
    }
  }
}

function insert(child, parent, anchor) {
  parent.insertBefore(child, anchor || null)
}

function remove(el) {
  const parent = el.parentNode;
  if (parent) {
    parent.removeChild(el)
  }
}

function setElementText(el: HTMLElement, text) {
  el.textContent =  text
}

const renderer: any = createRenderer({
  createElement,
  patchProps,
  insert,
  remove,
  setElementText,
})

export function createApp(...args) {
  return renderer.createApp(...args);
}

export * from '../runtime-core'