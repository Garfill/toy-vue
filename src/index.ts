//  toy-vue 入口文件
export * from './runtime-dom'
// export * from './reactive'

import * as runtimeDom from './runtime-dom'
import { registerRuntimeCompiler } from './runtime-dom'
import { baseCompile } from './compiler-core/src'

function compileToFunc(template) {
  const { code } = baseCompile(template)
  const render = new Function("Vue", code)(runtimeDom)
  return render
}

registerRuntimeCompiler(compileToFunc)