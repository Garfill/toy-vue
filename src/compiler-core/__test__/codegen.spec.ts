import { generate } from "../src/codegen";
import { baseParse } from "../src/parse";
import transform from "../src/transform";

test('string', () => {
  const ast: any = baseParse("hello");
  transform(ast, {})
  const { code } = generate(ast);
  expect(code).toMatchSnapshot();

})