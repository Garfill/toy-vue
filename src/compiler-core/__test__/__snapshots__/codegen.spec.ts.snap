// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen compbound 1`] = `
"const {toDisplayString: _toDisplayString, createElementVNode: _createElementVNode} = Vue
return function rendner(_ctx, _cache) {return _createElementVNode('div', null, 'hi, ' + _toDisplayString(_ctx.msg))}"
`;

exports[`codegen element 1`] = `
"const {createElementVNode: _createElementVNode} = Vue
return function rendner(_ctx, _cache) {return _createElementVNode('div', null, null)}"
`;

exports[`codegen interpolation module 1`] = `
"const {toDisplayString: _toDisplayString} = Vue
return function rendner(_ctx, _cache) {return _toDisplayString(_ctx.message)}"
`;

exports[`codegen string 1`] = `
"
return function rendner(_ctx, _cache) {return 'hello'}"
`;
